### PR TITLE
Add SerpApi Example

### PR DESCRIPTION
**Pull request by Zevleg**
Ticket [Implement Search API](https://app.clickup.com/t/85ztn9hk8)

Output result:

image

### CODE REVIEW

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                }
+            )
+        }
+    }
+
+    @Serializable
+    data class SearchData(
+        val q: String,
+        val location: String? = null,
+        val hl: String? = null,
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:38]: We can use here `@SerialName("hl") betterNameHere: String? = null` to have properties serialized as expected but still be able to make sense of them when we read them in kotlin.

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        val gl: String? = null,
+        val googleDomain: String = "google.com",
+        val apiKey: String
+    )
+
+    @Serializable data class SearchResults(@JsonNames("organic_results") val searchResult: List<SearchResult>)
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:44]: For better compatibility with Java, every class should be on its own file.

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                Json {
+                    encodeDefaults = false
+                    isLenient = true
+                    ignoreUnknownKeys = true
+                }
+            )
+        }
+    }
+
+    @Serializable
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:34]: Does this ever get serialized? It looks like the `SearchData` properties are used to construct a query string manually from its properties, which means this class does not need to be `@Serializable`

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        val response =
+            http.get(
+                "https://serpapi.com/search.json?q=${searchData.q}&location=${searchData.location}&hl=${searchData.hl}&gl=${searchData.gl}&google_domain=${searchData.googleDomain}&api_key=${searchData.apiKey}"
+            ) {
+                contentType(ContentType.Application.Json)
+            }
+
+        return if (response.status.isSuccess())
+            response.body<SearchResults>()
+        else throw SerpApiClientException(response.status, response.bodyAsText())
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:62]: We probably don't want to throw an exception here for some HTTP error codes. In some cases here we want to retry with some kind of exponential backoff. https://ktor.io/docs/client-retry.html#conditions

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        val title: String,
+        @JsonNames("snippet") val document: String,
+        val link: String
+    )
+
+    suspend fun search(searchData: SearchData): SearchResults {
+
+        val response =
+            http.get(
+                "https://serpapi.com/search.json?q=${searchData.q}&location=${searchData.location}&hl=${searchData.hl}&gl=${searchData.gl}&google_domain=${searchData.googleDomain}&api_key=${searchData.apiKey}"
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:55]: The `apiKey` should probably be in the `SerpApiClient` constructor as a nullable `String?`, if `null` is found, then we should load it from the environment with `getenv("SERP_API_KEY")` or similar. We don't want callers of this service to have to pass the key each time they create a request.

**Review by raulraja — CHANGES_REQUESTED** [review]
Looks great, but I think we should address the comments before merging so this class is consistent with how the other classes and services work with credentials, error handling, and names.

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt (hunk 5 above)
Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt]: Done

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt (hunk 4 above)
Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt]: Done

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt (hunk 1 above)
Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt]: Done

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt (hunk 3 above)
Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt]: Done

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import io.ktor.client.plugins.*
+import io.ktor.client.plugins.contentnegotiation.*
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+
+class SerpApiClient : AutoCloseable {
```
Comment (raulraja) [inline reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:15]: should implement `AutoClose by autoClose()` if we want it to close automatically with the `ai` block

**Review by raulraja — DISMISSED** [review]
looks great, added a final comment to ensure it autocloses

File: reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt (hunk 6 above)
Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt:15]: it has override close of autoClose() like in GcpClient.kt

```kotlin
override fun close() {
        http.close()
    }
``` 
it seems that implement is for abstract

image

Comment (Zevleg) [reply reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/serpapi/SerpApiClient.kt]: Done

**Review by raulraja — APPROVED** [review]
Looks great, thank you @Zevleg !

**Pull request merged**